code/campaignclassic/src/main/java/com/adobe/marketing/mobile/FilmstripCarouselTemplateNotificationBuilder.java

```diff
@@ -0,0 +1,314 @@
+
+        // log time needed to process the carousel images
+        final long imageProcessingElapsedTime =
+                System.currentTimeMillis() - imageProcessingStartTime;
+        Log.trace(
+                CampaignPushConstants.LOG_TAG,
+                SELF_TAG,
+                "Processed %d manual filmstrip carousel image(s) in %d milliseconds.",
+                downloadedImageUris.size(),
+                imageProcessingElapsedTime);
+
+        // fallback to a basic push template notification builder if less than 3 images were able
+        // to be downloaded
+        if (downloadedImageUris.size()
+                < CampaignPushConstants.DefaultValues.FILMSTRIP_CAROUSEL_MINIMUM_IMAGE_COUNT) {
+            Log.trace(
+                    CampaignPushConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Only %d image(s) for the manual filmstrip carousel notification were"
+                            + " downloaded. Building a basic push notification instead.",
+                    downloadedImageUris.size());
+            if (!CollectionUtils.isEmpty(downloadedImageUris)) {
+                // use the first downloaded image (if available) for the basic template notification
+                pushTemplate.modifyData(
+                        CampaignPushConstants.PushPayloadKeys.IMAGE_URL,
+                        downloadedImageUris.get(0));
+            }
+            final BasicPushTemplate basicPushTemplate =
+                    new BasicPushTemplate(pushTemplate.getData());
+            return BasicTemplateNotificationBuilder.construct(basicPushTemplate, context);
+        }
+
+        final String titleText = pushTemplate.getTitle();
+        final String smallBodyText = pushTemplate.getBody();
+        final String expandedBodyText = pushTemplate.getExpandedBodyText();
+        smallLayout.setTextViewText(R.id.notification_title, titleText);
+        smallLayout.setTextViewText(R.id.notification_body, smallBodyText);
+        expandedLayout.setTextViewText(R.id.notification_title, titleText);
+        expandedLayout.setTextViewText(R.id.notification_body_expanded, expandedBodyText);
+
+        // get all captions present then set center caption text
+        final String leftCaptionText = items.get(0).getCaptionText();
+        final String centerCaptionText = items.get(1).getCaptionText();
+        final String rightCaptionText = items.get(2).getCaptionText();
+        expandedLayout.setTextViewText(R.id.manual_carousel_filmstrip_caption, centerCaptionText);
+
+        // set the downloaded bitmaps in the filmstrip image views
+        expandedLayout.setImageViewBitmap(
+                R.id.manual_carousel_filmstrip_left, downloadedImages.get(0));
+        expandedLayout.setImageViewBitmap(
+                R.id.manual_carousel_filmstrip_center, downloadedImages.get(1));
+        expandedLayout.setImageViewBitmap(
+                R.id.manual_carousel_filmstrip_right, downloadedImages.get(2));
```

Comment:
Please refer to my comment on using arrays and indices to scale this to multiple images and combining notification building from payload and intent into one class. You can reuse a lot of logic in that way.